Opaque result types: dynamic replacement

aschwaighofer opened the pull request.

aschwaighofer committed:
Add a pass to specialize opaque type archetypes.

Clones functions containing opaque type archetypes replacing the opaque
type by a concrete type.

rdar://46140751

aschwaighofer committed:
Adjust test for 32bit platforms

aschwaighofer committed:
This test requires debug mode

aschwaighofer committed:
Remove code I added for a previous implementation

aschwaighofer committed:
Make opaque type descriptors dynamically replaceable

This is to support dynamic function replacement of functions with opaque
result type.

This approach requires that all state is thrown away (that could contain the
old returned type for an opaque type) between replacements.

rdar://48887938

aschwaighofer committed:
Fix tests

aschwaighofer committed:
[IRGen/TBDGen] Minor fixes for opaque result types.

Fixes rdar://problem/50001229.

aschwaighofer committed:
Opaque result types: Dynamic replacement fixes for computed properties

aschwaighofer commented:
@swift-ci Please test

slavapestov commented on lib/IRGen/Linking.cpp:
```diff
@@ -576,6 +592,10 @@ SILLinkage LinkEntity::getLinkage(ForDefinition_t forDefinition) const {
   case Kind::ProtocolRequirementsBaseDescriptor:
   case Kind::MethodLookupFunction:
   case Kind::OpaqueTypeDescriptor:
+  case Kind::OpaqueTypeDescriptorAccessor:
```
Do these really need to be public?

aschwaighofer replied:
Yed, they are what is getting replaced.

slavapestov commented on include/swift/IRGen/Linking.h:
```diff
@@ -199,6 +199,10 @@ class LinkEntity {
     /// The descriptor for an opaque type.
     /// The pointer is an OpaqueTypeDecl*.
     OpaqueTypeDescriptor,
+    OpaqueTypeDescriptorAccessor,
```
These should have comments

slavapestov reviewed (commented):
Sorry, I don’t know why some of my comments are not rendered properly.

slavapestov commented:
@aschwaighofer It looks like when dynamic replacement is off, we directly reference the opaque type descriptor and call a runtime function to get the concrete metadata. Is this correct? Then why do we need to emit an accessor function when dynamic replacement is on? Can we not dynamically replace a global variable, perhaps storing a pointer inside an indirect global variable?

Do the new symbols get emitted even when dynamic replacement is off?

Here's an example showing you can have a 'some P' type that's in parameter position, or nested inside a structural return type:

```
protocol P {}

protocol Q {
  associatedtype T

  func f() -> T
}

struct S : Q {
  func f() -> some P { return ... }
}

struct UsesSomePInFunnyWay {
  func blah(_ x: S.T) -> Array<S.T> { return [x] }
}
```

Perhaps dynamic replacement can't support this yet, but it should degrade gracefully instead of crashing in a cast.

slavapestov commented on lib/IRGen/GenDecl.cpp:
```diff
@@ -1341,6 +1338,36 @@ void IRGenerator::emitDynamicReplacements() {
+  SmallVector<OpaqueTypeArchetypeType *, 8> newFuncTypes;
+  SmallVector<OpaqueTypeArchetypeType *, 8> origFuncTypes;
+  for (auto *newFunc : DynamicReplacements) {
+    if (!newFunc->getLoweredFunctionType()->hasOpaqueArchetype())
```
This should be a structural walk over the SILFunctionType using `newFunc->getLoweredFunctionType().each()`

slavapestov commented on lib/SILOptimizer/Transforms/SpecializeOpaqueArchetypes.cpp:
```diff
@@ -0,0 +1,155 @@
+    auto argType = substConv.getSILArgumentType(argIdx);
+    if (argConv.isIndirectConvention() &&
+        argType.getASTType()->hasOpaqueArchetype() &&
+        !opd.get()->getType().getASTType()->hasOpaqueArchetype()) {
```
Maybe just check that the types differ? What if both have opaque types but they're different, because of various inlining and specialization passes?

slavapestov commented on lib/SILOptimizer/Transforms/SpecializeOpaqueArchetypes.cpp:
```diff
@@ -0,0 +1,155 @@
+      for (auto &inst : BB) {
+        auto *allocStack = dyn_cast<AllocStackInst>(&inst);
+        if (!allocStack ||
+            !allocStack->getElementType().is<OpaqueTypeArchetypeType>())
```
What if the element type contains an opaque type but its not at the top level?

slavapestov commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2093,6 +2093,94 @@ static Type getDynamicComparisonType(ValueDecl *value) {
   return interfaceType->removeArgumentLabels(numArgumentLabels);
 }
 
+static OpaqueTypeArchetypeType *getOpaqueResulType(const AbstractFunctionDecl *func) {
```
typo, should be "ResultType"

slavapestov commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -2093,6 +2093,94 @@ static Type getDynamicComparisonType(ValueDecl *value) {
+}
+
+static bool
+matchFunctionWithOpaqueResultType(const AbstractFunctionDecl *replacement,
```
Maybe instead this could be a flag for TypeBase::matches()? It really doesn't need to handle parameters and results differently, the only difference is when it finds an opaque type, instead of checking for equality it checks if they have the same constraints

aschwaighofer commented:
@slavapestov  

> It looks like when dynamic replacement is off, we directly reference the opaque type descriptor and call a runtime function to get the concrete metadata. Is this correct? Then why do we need to emit an accessor function when dynamic replacement is on? Can we not dynamically replace a global variable, perhaps storing a pointer inside an indirect global variable?

I reuse the mechanism of dynamically replacing a function which is already present. I don't want to invent a new mechanism/runtime support for a feature that will be supplanted by something else.

> Do the new symbols get emitted even when dynamic replacement is off?
No it should not. (see emitOpaqueTypeDescriptorAccessor) but there is a bug that I missed the same check (getAddressOfOpaqueTypeDescriptor) that I have to fix.

slavapestov commented:
@aschwaighofer Ok, if you're going to replace this soon anyway then the current approach is fine. Do you have a radar or JIRA for the followup change? I'm curious to learn more about it.

Do you mind still tidying up the walks over the type to fish out any opaque types it contains, though, as well as the function type replacement matching in TypeCheckAttr.cpp?

aschwaighofer commented:
Yes I will look at fixing nested opaque types.

aschwaighofer committed:
Only use opaque type accessors when the function is dynamic

aschwaighofer committed:
Add comments

aschwaighofer committed:
IRGen: Traverse function type to find opaque archetypes

aschwaighofer committed:
Add AllowCompatibleOpaqueTypeArchetypes as a Type.match flag

aschwaighofer committed:
OpaqueArchetypeSpecializer addressed review comments

aschwaighofer committed:
Attempt to fix test case on linux bot

aschwaighofer commented:
@swift-ci Please test

aschwaighofer committed:
IRGen: Cleanup through use of IRGenFunction

aschwaighofer committed:
Disable test case on linux for now

aschwaighofer commented:
@swift-ci Please test

slavapestov commented on lib/SILOptimizer/Transforms/SpecializeOpaqueArchetypes.cpp:
```diff
@@ -0,0 +1,154 @@
+    auto argConv = apply.getArgumentConvention(opd);
+    auto argIdx = apply.getCalleeArgIndex(opd);
+    auto argType = substConv.getSILArgumentType(argIdx);
+    if (argConv.isIndirectConvention() &&
```
Can an opaque archetype be loadable (hence the argument convention is direct) if its protocol is class-constrained? I think you want to handle that by emitting an unchecked_ref_cast

aschwaighofer replied:
Yes. Fixed. Test case added.

slavapestov commented on lib/SILOptimizer/Transforms/SpecializeOpaqueArchetypes.cpp:
```diff
@@ -0,0 +1,154 @@
+    bool foundOpaqueArchetype = false;
+    for (auto &BB : *getFunction()) {
+      for (auto &inst : BB) {
+        auto *allocStack = dyn_cast<AllocStackInst>(&inst);
```
Why not just look for any instruction containing an opaque type?

aschwaighofer replied:
Good point.

slavapestov commented on lib/SILOptimizer/Transforms/SpecializeOpaqueArchetypes.cpp:
```diff
@@ -0,0 +1,154 @@
+    return SuperTy::remapASTType(ty)
+        .subst(replacer, replacer,
+               SubstFlags::SubstituteOpaqueArchetypes |
+                   SubstFlags::AllowLoweredTypes)
```
Indentation

aschwaighofer replied:
clang-format does that.

slavapestov commented on lib/SILOptimizer/Transforms/SpecializeOpaqueArchetypes.cpp:
```diff
@@ -0,0 +1,154 @@
+    return SuperTy::remapConformance(type, conf)
+        .subst(type, replacer, replacer,
+               SubstFlags::SubstituteOpaqueArchetypes |
+                   SubstFlags::AllowLoweredTypes);
```
Indentation

aschwaighofer replied:
clang-format does that.

slavapestov commented on lib/SILOptimizer/Transforms/SpecializeOpaqueArchetypes.cpp:
```diff
@@ -0,0 +1,154 @@
+    ReplaceOpaqueTypesWithUnderlyingTypes replacer;
+    return SuperTy::remapSubstitutionMap(Subs).subst(
+        replacer, replacer,
+        SubstFlags::SubstituteOpaqueArchetypes | SubstFlags::AllowLoweredTypes);
```
This formatting is not consistent with the previous lines

aschwaighofer replied:
clang-format does that.

slavapestov commented on test/SILOptimizer/specialize_opaque_type_archetypes.swift:
```diff
@@ -0,0 +1,74 @@
+  return t
+}
+
+// CHECK-LABEL: sil @$s1A10testFooBaryyxAA1PRzlF : $@convention(thin) <T where T : P> (@in_guaranteed T) -> () {
```
Can you add a few more test cases?
- concrete type is address-only
- the protocol is class-constrained
- the opaque type is in another module (so it can't be specialized)
- assignment to a stored property whose type is opaque (use the typealias/associated type trick I showed you in the earlier comment). I believe this will crash right now because you have to insert a cast, just like you do for function arguments

slavapestov commented on lib/SILOptimizer/Transforms/SpecializeOpaqueArchetypes.cpp:
```diff
@@ -0,0 +1,154 @@
+
+    if (foundOpaqueArchetype) {
+      SubstitutionMap subsMap = getFunction()->getForwardingSubstitutionMap();
+      OpaqueSpecializerCloner s(subsMap, *getFunction());
```
If the opaque type is from a different module (and hence its concrete type is not known to us) this is going to do a lot of wasted work. Can you check for feasibility before cloning the entire function?

aschwaighofer replied:
Resilience is currently not respected by type replacement: https://github.com/apple/swift/blob/master/lib/AST/Type.cpp#L2459

omochi commented:
Why these two functionality are in one pull request?
In my understand, specialization of ORT is **compile time** optimization that replacement resilient representation to underlying concrete type to improve performance when caller and callee are in same module.
On the other hand, dynamic replacement is replacement function implementation in **loading time** which is very early timing with loading dynamic library in **runtime**.
Discussion is this. https://forums.swift.org/t/dynamic-method-replacement/16619

So compiler implementation of specialization is independent from dynamic replacement support for ORT.
Am I missing something?

This comment is not criticism about development process.
I am hobby compiler learner and just to know about internal source code structure.

aschwaighofer commented:
No good reason.

aschwaighofer committed:
Fixes for OpaqueArchetypeSpecializer

Look at all instructions whether we have to cast. Return/YieldInst must
not change the returned type.

aschwaighofer commented:
This does not yet have the fix for stores ...

aschwaighofer committed:
Remove SpecializeOpaqueArchetypes for now

aschwaighofer commented:
I am going to remove the OpaqueArchetypeSpecializer from this PR for now.

aschwaighofer commented:
@swift-ci Please test

aschwaighofer commented:
I have moved the specializer to a separate PR here: https://github.com/apple/swift/pull/24224

aschwaighofer commented:
@swift-ci Please test

omochi commented:
Thanks to response.

slavapestov commented:
Thanks for addressing the review feedback, LGTM!
